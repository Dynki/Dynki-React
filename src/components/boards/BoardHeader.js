import React from 'react';
import BoardForm from './BoardForm';
import BoardHeaderMenu from './BoardHeaderMenu';

const BoardHeader = ({ allowWrite, board, onUpdateBoard }) => {
    return (
        <section className="board__header">
            <BoardForm allowWrite={allowWrite} onUpdateBoard={onUpdateBoard} board={board}></BoardForm>
            <BoardHeaderMenu allowWrite={allowWrite} boardId={board.id}></BoardHeaderMenu>
        </section>
    )
}

export default BoardHeader;