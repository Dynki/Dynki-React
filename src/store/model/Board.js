import moment from 'moment';

export class Board {
    constructor(type, title, description, userInfo) {
        this.createdBy = userInfo.uid;
        this.createdDate = moment().toDate();
        this.description = description;
        this.title = title;
        this.isFolder = false;
        this.entities = [];
        this.columns = [{ title: 'Description', model: 'description', class: 'text' }];
        this.type = type;
    }

    
}
