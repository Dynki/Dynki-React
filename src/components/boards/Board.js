import React from 'react';
import { connect } from 'react-redux';
import { debounce } from 'lodash';
import { DragDropContext } from 'react-beautiful-dnd';

import BoardHeader from './BoardHeader';
import BoardDetail from './BoardDetail';
import BoardNewRow from './BoardNewRow';

import { updateBoard, newRow } from '../../store/actions/boardActions';

// Container for board components.
class Board extends React.Component {

    constructor(props) {
        super(props);

        // Debounce update methods, to stop spamming persitence logic.
        // The board updates are triggered by user input, specifically the key presses 
        // when editing existing or new row descriptions.
        // The debounce forces the code to wait until 1 second after the last key press, 
        // before it fires the persitence logic. 
        this.onUpdateBoard = debounce(this.onUpdateBoard, 1000);
        this.onNewRow = debounce(this.onNewRow, 1000);
    }

    // Triggered by child board detail component. 
    onUpdateBoard = (board) => {
        this.props.updateBoard(board);
    }

    // Triggered by child new row component.
    onNewRow = (description) => {
        this.props.newRow(description);
    }

    // Helper function for drag drop reordering or board rows.
    reorder = (list, startIndex, endIndex) => {
        const result = Array.from(list);
        const [removed] = result.splice(startIndex, 1);
        result.splice(endIndex, 0, removed);
    
        return result;
    };

    // Triggered when drag of board row have ended.
    // Reorders the entities (rows) within the board and calls logic to persist result.
    onDragEnd = (result) => {
        // dropped outside the list
        if (!result.destination) {
            return;
        }
             
        const newBoard = this.props.board;
        newBoard.entities = this.reorder(
            newBoard.entities,
            result.source.index,
            result.destination.index
        );

        this.onUpdateBoard(newBoard);
    }

    render() {
        console.log('Rerendering Board Component');

        return (
            (this.props.board ? 
                <section className="board">
                    <BoardHeader 
                        onUpdateBoard={this.onUpdateBoard}
                        board={this.props.board}>
                    </BoardHeader>
                    <DragDropContext onDragEnd={this.onDragEnd}>
                        <BoardDetail 
                            onUpdateBoard={this.onUpdateBoard}
                            board={this.props.board}>
                        </BoardDetail>
                    </DragDropContext>
                    <BoardNewRow onNewRow={this.onNewRow}></BoardNewRow>
                </section> : null 
            )
        )
    }
}

const mapStateToProps = (state) => {
    return{
      board: state.boards.currentBoard
    }
}

const mapDispatchToProps = (dispatch) => {
    return{
      updateBoard: (board) => dispatch(updateBoard(board)),
      newRow: (description) => dispatch(newRow(description))
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(Board);