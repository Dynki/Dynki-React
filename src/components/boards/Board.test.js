import React from 'react';
import { mount } from 'enzyme';
import { Board, mapDispatchToProps, mapStateToProps } from './Board';
import BoardHeader from './BoardHeader';

it('should render null when no boards', () => {
    const wrapper = mount(<Board />);
    expect(wrapper.find(BoardHeader)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
});

it('should render full component when boards are present', () => {
    const wrapper = mount(<Board board={true}/>);
    expect(wrapper.find(BoardHeader)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
});

