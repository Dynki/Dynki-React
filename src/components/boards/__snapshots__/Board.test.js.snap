// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render full component when boards are present 1`] = `
<section
  className="board"
>
  <BoardHeader
    board={true}
    onUpdateBoard={[Function]}
  />
  <DragDropContext
    onDragEnd={[Function]}
  >
    <BoardDetail
      board={true}
      onUpdateBoard={[Function]}
    />
  </DragDropContext>
  <BoardNewRow
    onNewRow={[Function]}
  />
</section>
`;

exports[`should render null when no boards 1`] = `""`;
